entrance_rando.py

```diff
@@ -0,0 +1,388 @@
+            # additional unplaced entrances into those regions)
+            if require_new_regions:
+                if all(e.connected_region in er_state.placed_regions for e in lookup):
+                    return False
+
+            # if we're on minimal accessibility and can guarantee the game is beatable,
+            # we can prevent a failure by bypassing future validity checks. this check may be
+            # expensive; fortunately we only have to do it once
+            if perform_validity_check and world.options.accessibility == Accessibility.option_minimal \
+                    and world.multiworld.has_beaten_game(er_state.collection_state, world.player):
+                # ensure that we have enough locations to place our progression
+                accessible_location_count = 0
+                prog_item_count = len(er_state.collection_state.prog_items[world.player])
+                # short-circuit location checking in this case
+                if prog_item_count == 0:
+                    return True
+                for region in er_state.placed_regions:
+                    for loc in region.locations:
+                        if loc.can_reach(er_state.collection_state):
+                            accessible_location_count += 1
+                            if accessible_location_count >= prog_item_count:
+                                perform_validity_check = False
+                                # pretend that this was successful to retry the current stage
+                                return True
+
+            unplaced_entrances = [entrance for region in world.multiworld.get_regions(world.player)
+                                  for entrance in region.entrances if not entrance.parent_region]
+            unplaced_exits = [exit_ for region in world.multiworld.get_regions(world.player)
+                              for exit_ in region.exits if not exit_.connected_region]
+            entrance_kind = "dead ends" if dead_end else "non-dead ends"
+            region_access_requirement = "requires" if require_new_regions else "does not require"
+            raise EntranceRandomizationError(
+                f"None of the available entrances are valid targets for the available exits.\n"
+                f"Randomization stage is placing {entrance_kind} and {region_access_requirement} "
+                f"new region access by default\n"
+                f"Placeable entrances: {lookup}\n"
+                f"Placeable exits: {placeable_exits}\n"
+                f"All unplaced entrances: {unplaced_entrances}\n"
+                f"All unplaced exits: {unplaced_exits}")
+
+    er_targets = [entrance for region in world.multiworld.get_regions(world.player)
+                  for entrance in region.entrances if not entrance.parent_region]
+    exits = [ex for region in world.multiworld.get_regions(world.player)
```

Comment:
It *might* be nice if you could optionally provide these yourself to do GER in "stages", but I also realize there's some problems with this approach since GER fundamentally happens from the Entry region in a front-fill way, and the "groups" feature seems to mostly take care of anything you'd want to do with "separate randomisation" anyway. Anything that doesn't fit it the world would probably just do itself, I suppose

Reply:
It actually used to be pass-it-yourself but every implementation ended up doing this so we  agreed it would be better to just take the work out. There are a few non-hacky ways the world could do it in stages, the main ones being to iteratively define the entrances and call GER repeatedly, or to use is_valid_source_transition and can_connect_to to filter out later stages until all of the earlier stages are complete. I suspect the vast majority of users will be quite happy with it as is though so treating such a use case as an edge case seems fine to me

Reply:
Hmmm but you could still let a world optionally provide these, right?
Like how `sweep_for_events` works:

https://github.com/ArchipelagoMW/Archipelago/blob/main/BaseClasses.py#L758-L760

Idk I don't think it's a huge deal, but it seems somewhat simple to add this in this case